src/audio/dp_queue.c

```diff
@@ -0,0 +1,391 @@
+					      void __sparse_cache *ptr, uint32_t size)
+{
+	/* wrap-around? */
+	if ((uintptr_t)ptr + size > (uintptr_t)dp_queue_buffer_end(dp_queue)) {
+		/* writeback till the end of circular buffer */
+		dcache_invalidate_region
+			(ptr, (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr);
+		size -= (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr;
+		ptr = (__sparse_force void __sparse_cache *)dp_queue->data_buffer;
+	}
+	/* invalidate rest of data */
+	dcache_invalidate_region(ptr, size);
+}
+
+static inline union lock_key dp_queue_lock(struct dp_queue *dp_queue)
+{
+	/* use cross-core spinlock in case of shared queue
+	 * When shared, dp_queue structure is located in not cached memory
+	 * as required by spinlock
+	 */
+	union lock_key key;
+
+	if (dp_queue_is_shared(dp_queue))
+		key.spinlock = k_spin_lock(&dp_queue->lock);
+	else
+		/* use faster irq_lock in case of not shared queue (located in cached mem) */
+		key.irq_lock = irq_lock();
+
+	return key;
+}
+
+static inline void dp_queue_unlock(struct dp_queue *dp_queue, union lock_key key)
+{
+	if (dp_queue_is_shared(dp_queue))
+		k_spin_unlock(&dp_queue->lock, key.spinlock);
+	else
+		irq_unlock(key.irq_lock);
+}
+
+struct sof_sink *dp_queue_get_sink(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_api;
+}
+
+struct sof_source *dp_queue_get_source(struct dp_queue *dp_queue)
+{
+	return &dp_queue->source_api;
+}
+
+static size_t dp_queue_get_free_size(struct sof_sink *sink)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	return dp_queue->free_space;
+}
+
+static int dp_queue_get_buffer(struct sof_sink *sink, size_t req_size,
+			       void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	/* dp_queue_get_free_size will return free size with adjustment for cacheline if needed */
+	if (req_size > dp_queue_get_free_size(sink))
+		return -ENODATA;
+
+	/* no need to lock, just reading data that may be modified by commit_buffer only
+	 *
+	 * note! a sparse warning will be generated here till
+	 * https://github.com/thesofproject/sof/issues/8006 is implemented
+	 */
+	*data_ptr = dp_queue->data_buffer + dp_queue->write_offset;
+	*buffer_start = dp_queue->data_buffer;
+	*buffer_size = dp_queue->data_buffer_size;
+
+	/* provided buffer is an empty space, the requester will perform write operations only
+	 * no need to invalidate cache - will be overwritten anyway
+	 */
+	return 0;
+}
+
+static int dp_queue_commit_buffer(struct sof_sink *sink, size_t commit_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	if (commit_size) {
+		union lock_key key = dp_queue_lock(dp_queue);
+
+		if (dp_queue_is_shared(dp_queue)) {
+			/* a shared queue. We need to go through committed cachelines one-by-one
+			 * and if the whole cacheline is committed - writeback cache
+			 * and mark data as available for reading
```

Comment:
>>What if the reader tries to read it on another core before the writer finishes the cache line?
Can't happen, the reader won't get this data - from reader point of view this data has not been produced yet. dp_queue->available_data won't be incremented till the cacheline is written back

As to using of a partially filled cacheline, as long as the reader won't try to modify data (it may result in cache writeback triggered by CPU and overwriting data being produced) it may be done, but for a price. The price will be one extra invalidation on consumer side and one extra writeback on producer (some of cache already written will have to be written back again, same applies to reading from cache).  I wanted to avoid it, but 1) according to my latest cache measurements and 2) the frame size is usually larger than cacheline and 3) dpQueue code will be simpler - it won't hurt, . Will give a try
also will add a clear comment to sink/source API that data from reading buffer can't be modified by the reader in any way

Reply:
> will add a clear comment to sink/source API that data from reading buffer can't be modified by the reader in any way

can we make all reader pointers `const`?